trustchain-http/src/verifier.rs

```diff
@@ -1,94 +1,400 @@
+                .await?,
+        )
+    }
+    /// Verifies verifiable credential.
+    async fn verify_credential<T: DIDResolver + Send + Sync>(
+        credential: &Credential,
+        root_event_time: Timestamp,
+        verifier: &IONVerifier<T>,
+    ) -> Result<(), TrustchainHTTPError> {
+        let verify_credential_result = credential.verify(None, verifier.resolver()).await;
+        if !verify_credential_result.errors.is_empty() {
+            return Err(TrustchainHTTPError::InvalidSignature);
+        }
+        match credential.get_issuer() {
+            Some(issuer) => Ok(verifier.verify(issuer, root_event_time).await.map(|_| ())?),
+            _ => Err(TrustchainHTTPError::NoCredentialIssuer),
+        }
+    }
 }
 
+/// Handler for verification of credentials and presentations.
 pub struct TrustchainVerifierHTTPHandler;
 
-impl TrustchainVerifierHTTP for TrustchainVerifierHTTPHandler {
-    fn generate_presentation_request(_presentation_id: &str) -> PresentationRequest {
-        todo!()
+impl TrustchainVerifierHTTP for TrustchainVerifierHTTPHandler {}
+
+/// Struct for deserializing credential and corresponding root event time.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PostVerifier {
+    pub presentation_or_credential: PresentationOrCredential,
+    pub root_event_time: Timestamp,
+}
+
+/// Enum for indicating whether verification information is a presentation or credential.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum PresentationOrCredential {
+    Presentation(Presentation),
+    Credential(Credential),
+}
+
+impl TrustchainVerifierHTTPHandler {
+    /// API endpoint taking the UUID of a presentation request.
+    pub async fn get_verifier(
+        Path(request_id): Path<String>,
+        State(app_state): State<Arc<AppState>>,
+    ) -> impl IntoResponse {
+        app_state
+            .presentation_requests
+            .get(&request_id)
+            .ok_or(TrustchainHTTPError::RequestDoesNotExist)
+            .map(|request| (StatusCode::OK, Json(request.to_owned())))
     }
+    /// Handler for presentation or credential received from POST.
+    pub async fn post_verifier(
+        Json(verification_info): Json<PostVerifier>,
+        app_state: Arc<AppState>,
+    ) -> impl IntoResponse {
+        let verification_info_json = serde_json::to_string_pretty(&verification_info)
+            .map_err(TrustchainHTTPError::FailedToDeserialize)?;
+        info!("Received verification information:\n{verification_info_json}",);
 
-    fn verify_presentation(_presentation: &Presentation) -> Result<(), PresentationError> {
-        todo!()
+        match verification_info.presentation_or_credential {
+            PresentationOrCredential::Presentation(ref presentation) => {
+                TrustchainVerifierHTTPHandler::verify_presentation(
+                    presentation,
+                    verification_info.root_event_time,
```

Comment:
Need to add `root_event_time` to the `app_state` so that this information is provided by the verifier instead of the holder.

Reply:
Completed